Tasks/Npm/npmtask.ts

```diff
@@ -1,27 +1,194 @@
+import url = require('url');
 import tl = require('vsts-task-lib/task');
+import trm = require('vsts-task-lib/toolrunner');
+
+interface EnvironmentDictionary { [key: string]: string; }
 
 tl.setResourcePath(path.join( __dirname, 'task.json'));
 
-var npm = tl.createToolRunner(tl.which('npm', true));
+executeTask();
+
+async function executeTask() {
+
+    var cwd = tl.getPathInput('cwd', true, false);
+    tl.mkdirP(cwd);
+    tl.cd(cwd);
+
+    var command = tl.getInput('command', true);
+    if (command.indexOf(' ') >= 0) {
+        tl.setResult(tl.TaskResult.Failed, tl.loc("InvalidCommand"));
+        return;
+    }
+
+    var npmRunner = tl.tool(tl.which('npm', true));
+    npmRunner.arg(command);
+    npmRunner.line(tl.getInput('arguments', false));
+
+
+    if(shouldUseDeprecatedTask()) {
+
+        // deprecated version of task, which just runs the npm command with NO auth support.
+        try{
+            var code : number = await npmRunner.exec();
+            tl.setResult(code, tl.loc('NpmReturnCode', code));
+        } catch (err) {
+            tl.debug('taskRunner fail');
+            tl.setResult(tl.TaskResult.Failed, tl.loc('NpmFailed', err.message));
+        }
+    } else {
+
+        // new task version with auth support
+        var npmrcPath: string = path.join(cwd, '.npmrc');
+        var tempNpmrcPath : string = getTempNpmrcPath();
+        var debugLog: boolean = true;
+        if (tl.getVariable('system.debug') != undefined && tl.getVariable('system.debug').toLowerCase() === 'false') {
+            debugLog = false;
+        }
+
+        try{
+            await runNpmAuthHelperAsync(getNpmAuthHelperRunner(npmrcPath, tempNpmrcPath, debugLog));
+
+            // set required environment variables for npm execution
+            let npmExecOptions = <trm.IExecOptions>{
+                env: process.env
+            };
+            npmExecOptions.env['npm_config_userconfig'] = tempNpmrcPath;
+            if(debugLog) {
+                npmExecOptions.env['npm_config_loglevel'] =  'verbose';
+            }
+
+            await tryRunNpmConfigAsync(getNpmConfigRunner(debugLog), npmExecOptions);
+            var code : number =  await runNpmCommandAsync(npmRunner, npmExecOptions);
+            cleanUpTempNpmrcPath(tempNpmrcPath);
+            tl.setResult(code, tl.loc('NpmReturnCode', code));
+        } catch (err) {
+            cleanUpTempNpmrcPath(tempNpmrcPath);
+            tl.setResult(tl.TaskResult.Failed, tl.loc('NpmFailed', err.message));
+        }
+    }
+}
+
+async function runNpmAuthHelperAsync(npmAuthRunner: trm.ToolRunner) : Promise<number> {
+    try{
+        var execOptions = <trm.IExecOptions>{
+            env: addBuildCredProviderEnv(process.env)
+        };
+
+        var code : number = await npmAuthRunner.exec(execOptions);
+        tl.debug('Auth helper exitted with code: ' + code);
+        return Q(code);
+    } catch (err) {
+        // ignore any auth failure and try to run the task.
```

Comment:
Agreed & changed. Although, this will cause failures for build defs which fails on warnings though. I think that's fine and user's attention should be drawn to any auth failure, even if their existing builds start to fail with the new task.